### PR TITLE
(CDAP-4264) Added system metadata for all CDAP entities

### PR DESCRIPTION
**Pull request by bdmogal**
(CDAP-4264) Add system metadata for all CDAP entities. 

(CDAP-4269) Added REST REST APIs to query system metadata by setting the 'scope' query parameter to 'system'.

(CDAP-4265) (CDAP-4266) (CDAP-4267) (CDAP-4268) Added the ability to add/update/delete system metadata for artifacts, apps, programs, datasets, streams and views.

Build: http://builds.cask.co/browse/CDAP-RBT592
Jiras:
[CDAP-4264](https://issues.cask.co/browse/CDAP-4264)
[CDAP-4265](https://issues.cask.co/browse/CDAP-4265)
[CDAP-4266](https://issues.cask.co/browse/CDAP-4266)
[CDAP-4267](https://issues.cask.co/browse/CDAP-4267)
[CDAP-4268](https://issues.cask.co/browse/CDAP-4268)
[CDAP-4269](https://issues.cask.co/browse/CDAP-4269)

This PR is based off of #4884, which has already been reviewed. Please only review the second commit.

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    return new String[]{programId.getType().getPrettyName(), getMode()};
+  }
+
+  public String getMode() {
+    String mode;
+    switch (programId.getType()) {
+      case MAPREDUCE:
+      case SPARK:
+      case WORKFLOW:
+      case CUSTOM_ACTION:
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java]: nit: just return instead of setting to local variable and `break`ing

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ViewSystemMetadataWriter.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    addSchema(properties, viewSpec.getFormat().getSchema());
+    return properties.build();
+  }
+
+  @Override
+  String[] getSystemTagsToAdd() {
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ViewSystemMetadataWriter.java]: Should the default implementation have `return new String[0]`?
That way, you only override it if you have tags that you want to add?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 3]
```diff
@@ -91,12 +99,35 @@ public void deleteAllModules(Id.Namespace namespaceId) throws DatasetManagementE
...
+    AbstractSystemMetadataWriter systemMetadataWriter =
+      new DatasetSystemMetadataWriter(metadataStore, dsInstantiatorFactory, datasetInstanceId, props, datasetTypeName);
+    systemMetadataWriter.write();
+  }
+
+  //TODO: Improve?
+  private boolean isUserDataset(Id.DatasetInstance datasetInstanceId) {
+    return !Id.Namespace.SYSTEM.equals(datasetInstanceId.getNamespace()) &&
+      !"system.queue.config".equals(datasetInstanceId.getId()) &&
+      !datasetInstanceId.getId().startsWith("system.sharded.queue") &&
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: I think if it starts with 'system.', that should be a sufficient check.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import co.cask.cdap.pipeline.AbstractStage;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.reflect.TypeToken;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Stage to write system metadata for the application.
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: for an application

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java [hunk 5]
```diff
@@ -234,39 +262,90 @@ public void testSearchOnTags() throws Exception {
...
+    result = results.get(0);
+    Assert.assertEquals(multiWordEntry, result);
+
+    // Search for it based split patterns to make sure nothing is matched
+    results =
+      dataset.search("ns1", "-", MetadataSearchTargetType.ALL);
+    // Assert check
+    Assert.assertEquals(0, results.size());
+
+    results =
```
Comment (anwar6953) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/dataset/MetadataDatasetTest.java]: why split the statement across two lines? It seems short enough.

File: cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java [hunk 6]
```diff
@@ -19,6 +19,8 @@
  * Supported types for metadata search.
  */
 public enum MetadataSearchTargetType {
+  // the custom values are required because these value match the entitiy-type stored as
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/metadata/MetadataSearchTargetType.java]: misspelling.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  private final MetadataStore metadataStore;
+
+  public SystemMetadataWriterStage(MetadataStore metadataStore) {
+    super(TypeToken.of(ApplicationWithPrograms.class));
+    this.metadataStore = metadataStore;
+  }
+
+  @Override
+  public void process(ApplicationWithPrograms input) throws Exception {
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: For Application it will be nice to 
- store the schedules the applications have.
- If the application is ETL Pipeline then the name of the plugins which it has

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java [hunk 8]
```diff
@@ -386,7 +391,15 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
     try {
       ArtifactClasses artifactClasses = inspectArtifact(artifactId, artifactFile, additionalPlugins, parentClassLoader);
       ArtifactMeta meta = new ArtifactMeta(artifactClasses, parentArtifacts, properties);
-      return artifactStore.write(artifactId, meta, Files.newInputStreamSupplier(artifactFile));
+      ArtifactDetail artifactDetail = artifactStore.write(artifactId, meta, Files.newInputStreamSupplier(artifactFile));
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: nit: metadata

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A class to write {@link MetadataScope#SYSTEM} metadata for an {@link Id.NamespacedId entity}.
+ */
+public abstract class AbstractSystemMetadataWriter {
+
+  public static final char COLON = ':';
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: We should re-use the constant MetadataDataset.KEYVALUE_SEPARATOR here too. 

Ideally, this constant should not be defined in MetadataDataset anymore. All the logic of what key-value is should be moved to MetadataStore. We can do that later.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: This is actually of type ``ArtifactMeta``, so I'll leave it as is

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    // add system metadata for apps
+    ApplicationSpecification appSpec = input.getSpecification();
+    AbstractSystemMetadataWriter appSystemMetadataWriter = new AppSystemMetadataWriter(metadataStore, input.getId(),
+                                                                                       appSpec);
+    appSystemMetadataWriter.write();
+
+    // add system metadata for programs
+    Id.Program programId;
+    List<AbstractSystemMetadataWriter> programSystemMetadataWriters = new ArrayList<>();
+    for (MapReduceSpecification mrSpec : appSpec.getMapReduce().values()) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: Instead of this highly similar for loop, create a method that look like this:

```java
private void addWriters(String programName, ProgramType programType, Iterable<? extends ProgramSpecification> specs, List<? extends AbstractSystemMetadataWriter> writers) {
  for (ProgramSpecification spec : specs) {
    Id.Program programId = Id.Program.from(programName, programType, spec.getName());
    writers.add(new ProgramSystemMetadataWriter(metaStore, programId, spec));
  }
}

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    for (SparkSpecification sparkSpec : appSpec.getSpark().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.SPARK, sparkSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, sparkSpec));
+    }
+    for (WorkerSpecification workerSpec : appSpec.getWorkers().values()) {
+      programId = Id.Program.from(input.getId(), ProgramType.WORKER, workerSpec.getName());
+      programSystemMetadataWriters.add(new ProgramSystemMetadataWriter(metadataStore, programId, workerSpec));
+    }
+
+    for (AbstractSystemMetadataWriter programSystemMetadataWriter : programSystemMetadataWriters) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: Why need to add all writers to a list and then loop it again to write? Can we just write in the for-loop above?

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: No reason, just wrote some lame code a month ago and didn't revisit. Fixing.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java [hunk 12]
```diff
@@ -386,7 +391,15 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
...
+      ArtifactDescriptor descriptor = artifactDetail.getDescriptor();
```
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java]: Seems like `descriptor` is unused?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java [hunk 13]
```diff
@@ -59,13 +64,17 @@ public boolean createOrUpdate(Id.Stream.View viewId, ViewSpecification spec) thr
       spec = new ViewSpecification(spec.getFormat(), naming.getTableName(viewId));
     }
     explore.enableExploreStream(viewId.getStream(), spec.getTableName(), spec.getFormat());
-    return store.createOrUpdate(viewId, spec);
+    boolean result = store.createOrUpdate(viewId, spec);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: If `createOrUpdate` returns false then the metadata should not be added, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import co.cask.cdap.data2.metadata.store.MetadataStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.metadata.MetadataScope;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A class to write {@link MetadataScope#SYSTEM} metadata for an {@link Id.NamespacedId entity}.
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: `entity` should be outside the {@link} tag.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java (hunk 13 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: So ``createOrUpdate`` returns ``false`` when the view already exists, and is updated. We want to update system metadata even when the view is updated, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+      return;
+    }
+
+    if (schema.isSimpleOrNullableSimple()) {
+      Schema.Type type = getSimpleType(schema);
+      propertiesToUpdate.put(SCHEMA_FIELD_PROPERTY_PREFIX, type.toString());
+    } else {
+      for (Schema.Field field : schema.getFields()) {
+        String fieldName = field.getName();
+        // TODO: What if field.getSchema() is not simple or nullable simple?
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: As we discussed, if we tokenize the search query then we could index all types of fields in schema. File a JIRA for this.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        String fieldType = getSimpleType(field.getSchema()).toString();
+        propertiesToUpdate.put(SCHEMA_FIELD_PROPERTY_PREFIX + COLON + fieldName, fieldName + COLON + fieldType);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: As we discussed, we need to store the entire schema as metadata. And break the schema down only for indexing.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java (hunk 14 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: Is `entity` link text? If so, ignore my comment.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    ArtifactClasses classes = artifactInfo.getClasses();
+    for (PluginClass pluginClass : classes.getPlugins()) {
+      String name = pluginClass.getName();
+      String type = pluginClass.getType();
+      // Need both name and type in the key because two plugins of different types could have the same name.
+      // However, the composite of name + type is guaranteed to be unique
+      properties.put(PLUGIN_KEY_PREFIX + COLON + name + COLON + type, name + COLON + type);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java]: The more I see the different formats of keys and values we index, the more I think we need to tokenize search query and then match.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java (hunk 14 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: yes, alias

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/DatasetSystemMetadataWriter.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+    Map<String, String> datasetProperties = dsProperties.getProperties();
+    String schemaStr = null;
+    if (datasetProperties.containsKey("schema")) {
+      schemaStr = datasetProperties.get("schema");
+    } else if (datasetProperties.containsKey("object.schema")) {
+      // If it is an ObjectMappedTable, the schema is in a property called 'object.schema'
+      schemaStr = datasetProperties.get("object.schema");
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/DatasetSystemMetadataWriter.java]: Don't we define constants for `schema` and `object.schema`?

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/DatasetSystemMetadataWriter.java]: yes, found them, coming in the next commit.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: just wanted to be as strict as possible. Anyway, we need to make this easier/robust somehow - https://issues.cask.co/browse/CDAP-4627

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: So a user cannot create a Dataset with a name beginning with `system`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java (hunk 7 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStage.java]: done

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java (hunk 13 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/view/ViewAdmin.java]: I see, `createOrUpdate` only tells whether view already existed or was created afresh. Ignore my comment then.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java (hunk 3 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: Its been a known issue since 2.8 - https://issues.cask.co/browse/CDAP-1587. We need to bump its priority

**Comment by poornachandra** [comment]
@bdmogal Overall this version of code looks good. I haven't looked at the tests yet. I'll do another pass tomorrow.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java (hunk 15 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: https://issues.cask.co/browse/CDAP-4628

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java (hunk 1 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java]: fixed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java (hunk 17 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java]: Also, how can this be searched?

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ArtifactSystemMetadataWriter.java]: The following queries should work:
1. ``<plugin name>``
2. ``<plugin type>``
3. ``<plugin name>:<plugin type>``
4. ``plugin:<plugin name>:<plugin type>``.

However, I'm probably missing a test for this, will add.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ViewSystemMetadataWriter.java (hunk 2 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ViewSystemMetadataWriter.java]: This is not a user-facing abstract class, its only to be extended for CDAP entities. I would prefer we override it for every entity, since there won't be a lot. When this PR was initially opened, there were a lot of empty properties/tags, but there are fewer such instances now.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 19]
```diff
@@ -308,7 +322,7 @@ public void testTags() throws Exception {
     );
     Assert.assertEquals(expected, searchTags);
     // test prefix search, should match stream and application
-    searchTags = searchMetadata(Id.Namespace.DEFAULT, "Wo*", MetadataSearchTargetType.ALL);
+    searchTags = searchMetadata(Id.Namespace.DEFAULT, "Wow*", MetadataSearchTargetType.ALL);
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Why change this ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 20]
```diff
@@ -552,6 +566,345 @@ public void testDeletedProgramHandlerStage() throws Exception {
...
+    manifest.getMainAttributes().put(ManifestFields.EXPORT_PACKAGE,
+                                     AllProgramsApp.AppPlugin.class.getPackage().getName());
+    Id.Artifact pluginArtifact = Id.Artifact.from(Id.Namespace.DEFAULT, "plugins", "1.0.0");
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(),
+                        addPluginArtifact(pluginArtifact, AllProgramsApp.AppPlugin.class, manifest,
+                                          new HashSet<ArtifactRange>()).getStatusLine().getStatusCode());
+    // search using artifact name
+    Set<MetadataSearchResultRecord> expected = ImmutableSet.of(new MetadataSearchResultRecord(pluginArtifact));
+    Set<MetadataSearchResultRecord> results = searchMetadata(Id.Namespace.DEFAULT, "plugins", null);
+    Assert.assertEquals(expected, results);
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: For plugin or Schema can we add a metadata with same key as name and make to test
one without plugin: which gives 2 results
another with plugin: which gives 1 to  make sure the scoping works ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 19 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Because it matches "Workflow" and "Worker" as well, after adding system tags

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    addPrograms(ProgramType.SERVICE, appSpec.getServices().values(), properties);
+    addPrograms(ProgramType.SPARK, appSpec.getSpark().values(), properties);
+    addPrograms(ProgramType.WORKER, appSpec.getWorkers().values(), properties);
+    addPrograms(ProgramType.WORKFLOW, appSpec.getWorkflows().values(), properties);
+  }
+
+  private void addPrograms(ProgramType programType, Iterable<? extends ProgramSpecification> specs,
+                           ImmutableMap.Builder<String, String> properties) {
+    for (ProgramSpecification spec : specs) {
+      properties.put(programType.getPrettyName() + MetadataDataset.KEYVALUE_SEPARATOR + spec.getName(),
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: so now with this user can scope to program type too ?
I think we have a lot of different scoping now beside schema.
@poornachandra : What's your thought on so many scopes ?
Even if we go with this we should for sure document this.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java (hunk 16 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AbstractSystemMetadataWriter.java]: Lets do the schema changes in a separate PR, this is becoming too big already.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+  @Override
+  String[] getSystemTagsToAdd() {
+    return new String[] {
+      appSpec.getName(),
+      appSpec.getArtifactId().getName()
+    };
+  }
+
+  private void addPrograms(ImmutableMap.Builder<String, String> properties) {
+    addPrograms(ProgramType.FLOW, appSpec.getFlows().values(), properties);
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: so now with this user can scope to program type too ?
I think we have a lot of different scoping now beside schema.
@poornachandra : What's your thought on so many scopes ?
Even if we go with this we should for sure document this.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: We don't really have an option, right? coz a program name is not unique by itself. It is unique only as a composite ``program-type+program-name``

The more I think of it, we should probably add something like a ``context`` for search. Search in the ``context`` of ``schema``, ``plugin``, ``programtype``, etc?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+      Schedule schedule = scheduleSpec.getSchedule();
+      String scheduleExpr;
+      if (schedule instanceof TimeSchedule) {
+        scheduleExpr = ((TimeSchedule) schedule).getCronEntry();
+      } else if (schedule instanceof StreamSizeSchedule) {
+        scheduleExpr = String.valueOf(((StreamSizeSchedule) schedule).getDataTriggerMB());
+      } else {
+        throw new IllegalStateException("Unknown schedule type: " + schedule.getClass().getName());
+      }
+      properties.put("schedule" + MetadataDataset.KEYVALUE_SEPARATOR + schedule.getName(),
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: Schedule expression can have special character like * and search expression is given through query param how will user search for this ? I don't think * is allowed in query param. Do we expect users to encode it ?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+                     scheduleExpr);
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: scheduleExp can have a lot of special character http://www.quartz-scheduler.org/documentation/quartz-1.x/tutorials/crontrigger
I don't think its a good idea to store or support query on it. Lets store the schedule description as value if one is provided.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/AppSystemMetadataWriter.java]: I agree. However, I think there is still value in storing the ``dataTriggerMB`` for a ``StreamSizeSchedule`` though. What do you think?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 20 above)
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Please review https://github.com/caskdata/cdap/commit/bcc4566f7cff28f6d06faeb0ce041bae6eb67ea3

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+  @Override
+  Map<String, String> getSystemPropertiesToAdd() {
+    return ImmutableMap.of();
+  }
+
+  @Override
+  String[] getSystemTagsToAdd() {
+    List<String> tags = ImmutableList.<String>builder()
+      .add(programId.getId())
+      .add(programId.getType().getPrettyName())
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java]: I am not sure how this will look like to user.

Say I have a something tagged as my_pretty_flow and this program will get tagged as flow
Now someone searches for 'flow' this will show up but it doesn't tell user why ?
If we store as key value then key can be programType  = flow which will given user context.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/system/ProgramSystemMetadataWriter.java]: if a user tags something as my_pretty_flow, then it should show up in the search results for 'flow', IMO. Having it as key = value is difficult, because the key will again have to delimited, because just program type may not be unique?

Even usability wise, when a user is on CDAP, tagging makes sense here, because everyone knows that a flow is a programType. Storing it against a key 'programType' is a little redundant.

**Comment by poornachandra** [comment]
LGTM

**Comment by rohitsinha54** [comment]
LGTM

**Comment by bdmogal** [comment]
For the last commit, I have tested all the relevant tests that it impacts. I do not want to wait for an entire build for this, because this PR blocks a couple of other fixes in metadata. Before the last commit, the successful build is http://builds.cask.co/browse/CDAP-RBT592-6. Have gotten the last commit reviewed. Going to squash and merge.

**Comment by bdmogal** [comment]
Build passed after the last commit too - http://builds.cask.co/browse/CDAP-RBT592-9
